File: js/src/gilded_rose.js
function Item(name, sell_in, quality) {
  this.name = name;
  this.sell_in = sell_in;
  this.quality = quality;
}

var items = [];

// Item Updater Class
class ItemUpdater {
  constructor(item) {
    this.item = item;
  }

  update() {
    if (this.item.name.includes("Conjured")) {
      this.updateConjured();
    } else {
      switch (this.item.name) {
        case "Aged Brie":
          this.updateAgedBrie();
          break;
        case "Backstage passes to a TAFKAL80ETC concert":
          this.updateBackstagePasses();
          break;
        case "Sulfuras, Hand of Ragnaros":
          return; // Sulfuras does not change
        default:
          this.updateNormalItem();
      }
    }

    if (this.item.name !== "Sulfuras, Hand of Ragnaros") {
      this.item.sell_in -= 1;
    }

    if (this.item.sell_in < 0) {
      this.handleExpiredItem();
    }
  }

  increaseQuality(amount = 1) {
    this.item.quality = Math.min(50, this.item.quality + amount);
  }

  degradeQuality(amount = 1) {
    this.item.quality = Math.max(0, this.item.quality - amount);
  }

  updateAgedBrie() {
    this.increaseQuality();
  }

  updateBackstagePasses() {
    if (this.item.sell_in > 10) {
      this.increaseQuality();
    } else if (this.item.sell_in > 5) {
      this.increaseQuality(2);
    } else if (this.item.sell_in > 0) {
      this.increaseQuality(3);
    } else {
      this.item.quality = 0; // Expired backstage passes
    }
  }

  updateNormalItem() {
    this.degradeQuality();
  }

  updateConjured() {
    this.degradeQuality(2);
  }

  handleExpiredItem() {
    if (this.item.name === "Aged Brie") {
      this.increaseQuality();
    } else if (this.item.name === "Backstage passes to a TAFKAL80ETC concert") {
      this.item.quality = 0;
    } else if (this.item.name.includes("Conjured")) {
      this.degradeQuality(2);
    } else {
      this.degradeQuality();
    }
  }
}

// Refactored update_quality function
function update_quality() {
  items.forEach((item) => new ItemUpdater(item).update());
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-
class GildedRose:
    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            self.get_updater(item).update()

    def get_updater(self, item):
        """Returns the appropriate updater class for the item"""
        if item.name == "Aged Brie":
            return AgedBrieUpdater(item)
        elif item.name == "Backstage passes to a TAFKAL80ETC concert":
            return BackstagePassUpdater(item)
        elif item.name == "Sulfuras, Hand of Ragnaros":
            return SulfurasUpdater(item)
        elif "Conjured" in item.name:
            return ConjuredItemUpdater(item)
        else:
            return NormalItemUpdater(item)


class ItemUpdater:
    """Base class for updating item quality and sell-in values"""

    def __init__(self, item):
        self.item = item

    def update(self):
        self.update_quality()
        self.update_sell_in()
        if self.item.sell_in < 0:
            self.handle_expired()

    def update_quality(self):
        """Default behavior for normal items"""
        self.decrease_quality()

    def update_sell_in(self):
        """All items except Sulfuras decrease sell_in"""
        self.item.sell_in -= 1

    def handle_expired(self):
        """Handles quality changes when sell_in < 0"""
        self.decrease_quality()

    def increase_quality(self, amount=1):
        """Increases quality but caps it at 50"""
        self.item.quality = min(50, self.item.quality + amount)

    def decrease_quality(self, amount=1):
        """Decreases quality but ensures it doesn't go below 0"""
        self.item.quality = max(0, self.item.quality - amount)


class AgedBrieUpdater(ItemUpdater):
    def update_quality(self):
        self.increase_quality()

    def handle_expired(self):
        self.increase_quality()


class BackstagePassUpdater(ItemUpdater):
    def update_quality(self):
        if self.item.sell_in > 10:
            self.increase_quality()
        elif self.item.sell_in > 5:
            self.increase_quality(2)
        elif self.item.sell_in > 0:
            self.increase_quality(3)
        else:
            self.item.quality = 0  # Expired passes have 0 quality


class SulfurasUpdater(ItemUpdater):
    def update_quality(self):
        pass  # Sulfuras never changes

    def update_sell_in(self):
        pass  # Sulfuras never changes


class ConjuredItemUpdater(ItemUpdater):
    def update_quality(self):
        self.decrease_quality(2)  # Conjured items degrade twice as fast

    def handle_expired(self):
        self.decrease_quality(2)


class NormalItemUpdater(ItemUpdater):
    pass  # Uses default behavior from ItemUpdater


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return f"{self.name}, {self.sell_in}, {self.quality}"
